In [1]:
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report

##### Query for downloading the data from Google Cloud BIGQUERY

##### The dataset was downloaded as csv and later uploaded in Visual Studio where futher processing was performed

In [2]:
#importing dataset by reading csv file (The data was downloaded from Google Cloud using BIGQUERY)
df=pd.read_csv('/root/code/thesis/codeFolder/LatestDataInUse/csv/7learnings.csv')

/tmp/ipykernel_572666/1797354241.py:2: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('/root/code/thesis/codeFolder/LatestDataInUse/csv/7learnings.csv')


In [3]:
df.head()

station_number  wban_number  year  month  day  mean_temp  \
0          725940        99999  2005      1   25  51.299999   
1          725940        99999  2005      3    7  51.500000   
2          725940        99999  2005      5   11  51.000000   
3          725869        99999  2005      1   27  30.200001   
4          725827        99999  2005     10   23  55.000000   

   num_mean_temp_samples  mean_dew_point  num_mean_dew_point_samples  \
0                      4       45.299999                         4.0   
1                      4       48.000000                         4.0   
2                      4       45.700001                         4.0   
3                      5       26.600000                         5.0   
4                      5       28.799999                         5.0   

   mean_sealevel_pressure  ...  min_temperature_explicit  total_precipitation  \
0             1013.500000  ...                       NaN                 0.00   
1             1025.000000  ...                       NaN                 0.00   
2             1021.799988  ...                       NaN                 0.01   
3                     NaN  ...                       NaN                 0.00   
4             1015.599976  ...                       NaN                 0.00   

   snow_depth    fog   rain   snow   hail  thunder  tornado  formatted_date  
0         NaN  False  False  False  False    False    False      2005-01-25  
1         NaN  False  False  False  False    False    False      2005-03-07  
2         NaN  False  False  False  False    False    False      2005-05-11  
3         NaN  False  False  False  False    False    False      2005-01-27  
4         NaN  False  False  False  False    False    False      2005-10-23  

[5 rows x 32 columns]

# Data Preprocessing

In [4]:
# Filtering stations 725300 to 725330 that have information from 2005 till 2009.
df = df[(df['station_number'] >= 725300) & (df['station_number'] <= 725330)]

# Specify the columns to drop 
# Some of these column like were irrelevant, some coulumns had missing values
columns_to_drop = ['min_temperature', 'min_temperature_explicit','mean_station_pressure','mean_sealevel_pressure', 'num_mean_station_pressure_samples','year','month','day','snow_depth', 'num_mean_sealevel_pressure_samples', 'wban_number', 'num_mean_temp_samples','num_mean_dew_point_samples', 'num_mean_visibility_samples' ,'max_sustained_wind_speed','max_gust_wind_speed','max_temperature_explicit', 'num_mean_wind_speed_samples','tornado','max_temperature']

# Drop the specified columns
df_dropped = df.drop(columns=columns_to_drop)


In [5]:
# Checking columns with NAN values
columns_with_nan = df_dropped.columns[df_dropped.isna().any()].tolist()

print(columns_with_nan)


['mean_wind_speed', 'total_precipitation']


##### Strategies for dealing with missing values

In [6]:
#### Strategy 1: Filling the missing values with mean values
#data.fillna(data.mean(), inplace=True)

### Strategy 2: Interpolating the missing values
#df_dropped['mean_wind_speed'] = df_dropped['mean_wind_speed'].interpolate()
#df_dropped['total_precipitation'] = df_dropped['total_precipitation'].interpolate()
#df_dropped['mean_sealevel_pressure'] = df_dropped['mean_sealevel_pressure'].interpolate()

### Strategy 3: Droping the missing rows
df_dropped=df_dropped.dropna()

#### Droping the rows with missing values reduces data size but other strategies are more likely to compromise data quality

In [7]:
# Find columns with NaN values
columns_with_nan = df_dropped.columns[df_dropped.isna().any()].tolist()

print(columns_with_nan)


[]


In [8]:
# Sorting the dataframe with respect to data and resting the dataframe index
final_df = df_dropped.sort_values(by='formatted_date') 
final_df = final_df.reset_index(drop=True)

In [9]:
final_df.head(5)

station_number  mean_temp  mean_dew_point  mean_visibility  \
0          725314  60.799999       57.599998              8.9   
1          725300  32.299999       29.400000              7.7   
2          725330  24.799999       20.400000              9.7   
3          725305  16.700001       13.100000              8.4   
4          725316  28.600000       25.500000              7.3   

   mean_wind_speed  total_precipitation    fog   rain   snow   hail  thunder  \
0              4.8                 0.40  False  False  False  False    False   
1              9.3                 0.34   True   True   True   True     True   
2              7.7                 0.03  False  False  False  False    False   
3              6.9                 0.01  False  False  False  False    False   
4              5.4                 0.10   True   True   True   True     True   

  formatted_date  
0     2005-01-02  
1     2005-01-04  
2     2005-01-07  
3     2005-01-07  
4     2005-01-08

In [10]:
final_df.columns

Index(['station_number', 'mean_temp', 'mean_dew_point', 'mean_visibility',
       'mean_wind_speed', 'total_precipitation', 'fog', 'rain', 'snow', 'hail',
       'thunder', 'formatted_date'],
      dtype='object')

In [11]:
# Converting the categorical values to numerical values
final_df['fog'] = final_df['fog'].astype(int)
final_df['rain'] = final_df['rain'].astype(int)
final_df['hail'] = final_df['hail'].astype(int)
final_df['thunder'] = final_df['thunder'].astype(int)
final_df['snow'] = final_df['snow'].astype(int)

In [12]:
final_df.head()

station_number  mean_temp  mean_dew_point  mean_visibility  \
0          725314  60.799999       57.599998              8.9   
1          725300  32.299999       29.400000              7.7   
2          725330  24.799999       20.400000              9.7   
3          725305  16.700001       13.100000              8.4   
4          725316  28.600000       25.500000              7.3   

   mean_wind_speed  total_precipitation  fog  rain  snow  hail  thunder  \
0              4.8                 0.40    0     0     0     0        0   
1              9.3                 0.34    1     1     1     1        1   
2              7.7                 0.03    0     0     0     0        0   
3              6.9                 0.01    0     0     0     0        0   
4              5.4                 0.10    1     1     1     1        1   

  formatted_date  
0     2005-01-02  
1     2005-01-04  
2     2005-01-07  
3     2005-01-07  
4     2005-01-08

In [13]:
final_df.shape

(1524, 12)

In [14]:
np.unique(final_df['station_number'])

array([725300, 725305, 725314, 725315, 725316, 725317, 725320, 725326,
       725327, 725330])

### The dataset has huge class imbalance with 1350 values representing class 0 "no snow" and only 174 values prepresenting class 1 "snow"

In [15]:
final_df['snow'].value_counts()

0    1350
1     174
Name: snow, dtype: int64

# Splitting the data into train and Test Split

In [16]:

# Create features and labels
features = ['station_number', 'mean_temp', 'mean_dew_point', 'mean_visibility',
            'mean_wind_speed', 'total_precipitation', 'fog', 'rain', 'hail',
            'thunder']
X = final_df[features]

y = final_df['snow'].shift(-1).fillna(0).astype(int)  # Forecasting snow on the next day

# Splitting the data into train test split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Converting the labels to numpy arrays
y_train = y_train.values
y_test = y_test.values



In [17]:
# Standardizing the features to ensure equal weitage of each feature
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


# Model

In [18]:
# Create a custom dataset class
class ClimateDataset(Dataset):
    def __init__(self, features, labels):
        self.features = torch.tensor(features, dtype=torch.float32)
        self.labels = torch.tensor(labels, dtype=torch.float32)
    
    def __len__(self):
        return len(self.features)
    
    def __getitem__(self, idx):
        return self.features[idx], self.labels[idx]


In [19]:
# Define the neural network model
class SnowForecastNN(nn.Module):
    def __init__(self, input_size):
        super(SnowForecastNN, self).__init__()
        self.fc1 = nn.Linear(input_size, 64)
        self.fc2 = nn.Linear(64, 32)
        self.fc3 = nn.Linear(32, 1)
        self.relu = nn.ReLU()
        self.sigmoid = nn.Sigmoid()
    
    def forward(self, x):
        x = self.relu(self.fc1(x))
        x = self.relu(self.fc2(x))
        x = self.sigmoid(self.fc3(x))
        return x

# Training and Testing

In [20]:
train_dataset = ClimateDataset(X_train, y_train)
test_dataset = ClimateDataset(X_test, y_test)

train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False)


# Initialize the model, loss function, and optimizer
model = SnowForecastNN(input_size=X_train.shape[1])
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Training the model
num_epochs = 100
for epoch in range(num_epochs):
    model.train()
    running_loss = 0.0
    correct = 0
    total = 0
    for inputs, labels in train_loader:
        labels = labels.view(-1, 1)
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
        running_loss += loss.item()
        predicted = (outputs > 0.5).float()
        total += labels.size(0)
        correct += (predicted == labels).sum().item()
    
    train_loss = running_loss / len(train_loader)
    train_accuracy = 100 * correct / total
    
    # Evaluate the model
    model.eval()
    test_loss = 0.0
    correct = 0
    total = 0
    all_labels = []
    all_predictions = []
    with torch.no_grad():
        for inputs, labels in test_loader:
            labels = labels.view(-1, 1)
            outputs = model(inputs)
            loss = criterion(outputs, labels)
            test_loss += loss.item()
            predicted = (outputs > 0.5).float()
            total += labels.size(0)
            correct += (predicted == labels).sum().item()
            all_labels.extend(labels.cpu().numpy())
            all_predictions.extend(predicted.cpu().numpy())
    
    test_loss /= len(test_loader)
    test_accuracy = 100 * correct / total
    
    print(f'Epoch {epoch+1}/{num_epochs}, Train Loss: {train_loss:.4f}, Train Accuracy: {train_accuracy:.2f}%, Test Loss: {test_loss:.4f}, Test Accuracy: {test_accuracy:.2f}%')
    

Epoch 1/100, Train Loss: 0.5706, Train Accuracy: 78.67%, Test Loss: 0.4306, Test Accuracy: 88.52%
Epoch 2/100, Train Loss: 0.3737, Train Accuracy: 88.60%, Test Loss: 0.3491, Test Accuracy: 88.52%
Epoch 3/100, Train Loss: 0.3481, Train Accuracy: 88.60%, Test Loss: 0.3443, Test Accuracy: 88.52%
Epoch 4/100, Train Loss: 0.3425, Train Accuracy: 88.60%, Test Loss: 0.3420, Test Accuracy: 88.52%
Epoch 5/100, Train Loss: 0.3414, Train Accuracy: 88.60%, Test Loss: 0.3420, Test Accuracy: 88.52%
Epoch 6/100, Train Loss: 0.3392, Train Accuracy: 88.60%, Test Loss: 0.3416, Test Accuracy: 88.52%


Epoch 7/100, Train Loss: 0.3365, Train Accuracy: 88.60%, Test Loss: 0.3411, Test Accuracy: 88.52%
Epoch 8/100, Train Loss: 0.3563, Train Accuracy: 88.60%, Test Loss: 0.3389, Test Accuracy: 88.52%
Epoch 9/100, Train Loss: 0.3360, Train Accuracy: 88.60%, Test Loss: 0.3401, Test Accuracy: 88.52%
Epoch 10/100, Train Loss: 0.3334, Train Accuracy: 88.60%, Test Loss: 0.3397, Test Accuracy: 88.52%
Epoch 11/100, Train Loss: 0.3323, Train Accuracy: 88.60%, Test Loss: 0.3380, Test Accuracy: 88.52%
Epoch 12/100, Train Loss: 0.3327, Train Accuracy: 88.60%, Test Loss: 0.3394, Test Accuracy: 88.52%
Epoch 13/100, Train Loss: 0.3319, Train Accuracy: 88.60%, Test Loss: 0.3371, Test Accuracy: 88.52%
Epoch 14/100, Train Loss: 0.3511, Train Accuracy: 88.60%, Test Loss: 0.3362, Test Accuracy: 88.52%
Epoch 15/100, Train Loss: 0.3513, Train Accuracy: 88.60%, Test Loss: 0.3368, Test Accuracy: 88.52%
Epoch 16/100, Train Loss: 0.3467, Train Accuracy: 88.60%, Test Loss: 0.3364, Test Accuracy: 88.52%
Epoch 17/100,

In [21]:
# Print classification report
print("Classification Report:")
print(classification_report(all_labels, all_predictions, target_names=['No Snow', 'Snow']))

Classification Report:
              precision    recall  f1-score   support

     No Snow       0.88      0.98      0.93       270
        Snow       0.00      0.00      0.00        35

    accuracy                           0.87       305
   macro avg       0.44      0.49      0.46       305
weighted avg       0.78      0.87      0.82       305



### Although 87% test accuracy is achieved but the model is predicting mostly the majority class. In such cases techniques like stratifies group k-fold cross-validation and weighting random sampling can be used to mitigate the class imabalce